staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,460 @@
+use std::convert::TryInto;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::Map;
+use near_sdk::json_types::{Base58PublicKey, U128};
+use near_sdk::{env, near_bindgen, AccountId, Balance, EpochHeight, Promise, PublicKey};
+
+use uint::construct_uint;
+
+const PING_GAS: u64 = 30_000_000_000_000;
+const REFRESH_STAKE_GAS: u64 = 30_000_000_000_000;
+
+construct_uint! {
+    /// 256-bit unsigned integer.
+    pub struct U256(4);
+}
+
+#[cfg(test)]
+mod test_utils;
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+#[derive(BorshDeserialize, BorshSerialize, Debug, PartialEq)]
+pub struct Account {
+    pub unstaked: Balance,
+    pub staked_shares: Balance,
+    pub unstaked_available_epoch_height: EpochHeight,
+}
+
+impl Default for Account {
+    fn default() -> Self {
+        Self {
+            unstaked: 0,
+            staked_shares: 0,
+            unstaked_available_epoch_height: 0,
+        }
+    }
+}
+
+const EPOCHS_TOWARDS_REWARD: EpochHeight = 3;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct StakingContract {
+    pub owner_id: AccountId,
+    pub stake_public_key: PublicKey,
+    pub last_epoch_height: EpochHeight,
+    pub last_locked_account_balance: Balance,
+    pub last_account_balance: Balance,
+    pub total_staked_shares: Balance,
+    pub total_staked_balance: Balance,
+    pub accounts: Map<AccountId, Account>,
+}
+
+impl Default for StakingContract {
+    fn default() -> Self {
+        env::panic(b"Staking contract should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl StakingContract {
+    /// Call to initialize the contract.
+    /// Specify which account can change the staking key and the initial staking key with ED25519 curve.
+    #[init]
+    pub fn new(owner_id: AccountId, stake_public_key: Base58PublicKey) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            owner_id,
+            stake_public_key: stake_public_key
+                .try_into()
+                .expect("invalid staking public key"),
+            last_epoch_height: env::epoch_height(),
+            last_locked_account_balance: 0,
+            last_account_balance: env::account_balance(),
+            total_staked_balance: 0,
+            total_staked_shares: 0,
+            accounts: Map::new(b"u".to_vec()),
+        }
+    }
+
+    /// Call to update state after epoch switched.
+    pub fn ping(&mut self) {
+        // Checking if we need there are rewards to distribute.
+        let epoch_height = env::epoch_height();
+        if self.last_epoch_height == epoch_height {
+            return;
+        }
+        self.last_epoch_height = epoch_height;
+
+        let total_balance =
+            env::account_locked_balance() + env::account_balance() - env::attached_deposit();
```

Comment:
why there is attached deposit?
this function is not payable...

Reply:
I guess it's when this function is called from deposit function.

Reply:
Yep, in case it's called from deposit, the `account_balance` include the deposited amount